forward/testing/packet_synthesis.lua

```diff
@@ -0,0 +1,168 @@
+local GRE = require("lib.protocol.gre")
+local Datagram = require("lib.protocol.datagram")
+local Ethernet = require("lib.protocol.ethernet")
+local IPV4 = require("lib.protocol.ipv4")
+local TCP = require("lib.protocol.tcp")
+local ffi = require("ffi")
+
+local networking_magic_numbers = require("networking_magic_numbers")
+
+local function make_payload(len)
+   local buf = ffi.new('char[?]', len)
+   for i=0,len-1 do
+      buf[i] = (i + 128) % 256
+   end
+   return ffi.string(buf, len)
+end
+
+-- Returns an array of payload fragments.
+local function split_payload(payload, payload_len, fragment_len)
+   -- add fragment_payload_len-1 to handle case when payload_len is divisible by fragment_payload_len
+   -- want to compute (payload_len + fragment_len - 1) // fragment_payload_len but lua5.1 has no integer division
+   local dividend = payload_len + fragment_len - 1
+   local num_fragments = (dividend - dividend % fragment_len) / fragment_len
+   local fragments = {n = num_fragments}
+   for i=1,num_fragments do
+      local curr_fragment_len = fragment_len
+      if i == num_fragments then
+         curr_fragment_len = payload_len - fragment_len * (num_fragments-1)
+      end
+      local fragment = string.sub(payload, (i-1) * fragment_len + 1, math.min(i * fragment_len, payload_len))
+      fragments[i] = fragment
+   end
+   return fragments
+end
+
+-- Arguments:
+-- src_mac, dst_mac (binary) -- source and destination MAC addresses;
+--                              should be the addresses of the router
+--                              (or last switch) and spike respectively
+-- src_addr, dst_addr (binary) -- source and destination IP addresses;
+--                                should be the addresses of the router
+--                                and spike respectively
```

Comment:
The source address is not the address of the router, and the destination address is not the address of spike, I think.  The source address would usually be an external client's, and the destination address would be the VIP for a service.